docker/src/main/java/brooklyn/networking/sdn/SdnProviderImpl.java

```diff
@@ -0,0 +1,198 @@
+        }
+
+        setAttribute(SDN_AGENTS, agents);
+
+        setAttribute(ALLOCATED_IPS, 0);
+        setAttribute(ALLOCATED_ADDRESSES, Maps.<String, InetAddress>newConcurrentMap());
+
+        setAttribute(ALLOCATED_NETWORKS, 0);
+        setAttribute(NETWORKS, Maps.<String, Cidr>newConcurrentMap());
+        setAttribute(NETWORK_ALLOCATIONS, Maps.<String, Integer>newConcurrentMap());
+        setAttribute(CONTAINER_ADDRESSES, Maps.<String, InetAddress>newConcurrentMap());
+    }
+
+    @Override
+    public InetAddress getNextAgentAddress(String agentId) {
+        synchronized (addressMutex) {
+            Cidr cidr = getConfig(AGENT_CIDR);
+            Integer allocated = getAttribute(ALLOCATED_IPS);
+            InetAddress next = cidr.addressAtOffset(allocated + 1);
+            setAttribute(ALLOCATED_IPS, allocated + 1);
+            Map<String, InetAddress> addresses = getAttribute(ALLOCATED_ADDRESSES);
+            addresses.put(agentId, next);
+            setAttribute(ALLOCATED_ADDRESSES, addresses);
+            return next;
+        }
+    }
+
+    @Override
+    public InetAddress getNextContainerAddress(String subnetId) {
+        synchronized (addressMutex) {
+            Cidr cidr = getAttribute(NETWORKS).get(subnetId);
+            Map<String, Integer> allocations = getAttribute(NETWORK_ALLOCATIONS);
+            Integer allocated = allocations.get(subnetId);
+            if (allocated == null) allocated = 1;
+            InetAddress next = cidr.addressAtOffset(allocated + 1);
+            allocations.put(subnetId, allocated + 1);
+            setAttribute(NETWORK_ALLOCATIONS, allocations);
+            return next;
+        }
+    }
+
+    @Override
+    public Cidr getSubnet(String subnetId, String subnetName) {
+        synchronized (addressMutex) {
+            Map<String, Cidr> networks = getAttribute(NETWORKS);
+            if (networks.containsKey(subnetId)) return networks.get(subnetId);
+
+            Cidr networkCidr = getConfig(CONTAINER_NETWORK_CIDR);
+            Integer networkSize = getConfig(CONTAINER_NETWORK_SIZE);
+            Integer allocated = getAttribute(ALLOCATED_NETWORKS);
+
+            InetAddress baseAddress = networkCidr.addressAtOffset(allocated * (1 << (32 - networkSize)));
+            Cidr subnetCidr = new Cidr(baseAddress.getHostAddress() + "/" + networkSize);
+            LOG.debug("Allocated {} as {} from {} for subnet #{}", new Object[] { subnetCidr, subnetId, networkCidr, allocated });
+
+            networks.put(subnetId, subnetCidr);
+            setAttribute(ALLOCATED_NETWORKS, allocated + 1);
+            setAttribute(NETWORKS, networks);
+
+            return subnetCidr;
+        }
+    }
+
+    @Override
+    public DynamicCluster getDockerHostCluster() {
+        return getConfig(DOCKER_INFRASTRUCTURE).getAttribute(DockerInfrastructure.DOCKER_HOST_CLUSTER);
+    }
+
+    @Override
+    public Group getAgents() { return getAttribute(SDN_AGENTS); }
```

Comment:
Not necessary, because it isn't the agent group entity that is being protected by the mutex